From 57c638b27451c55dd226f1c5f9ba82c2a0db92df Mon Sep 17 00:00:00 2001
From: jatin <jatinchowdhury18@gmail.com>
Date: Wed, 2 Aug 2023 13:41:21 -0700
Subject: [PATCH] Check if compiler supports SIMD flags before applying

---
 modules/cmake/RuntimeSIMDLib.cmake | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/modules/cmake/RuntimeSIMDLib.cmake b/modules/cmake/RuntimeSIMDLib.cmake
index b40d115d..63239f4c 100644
--- a/modules/cmake/RuntimeSIMDLib.cmake
+++ b/modules/cmake/RuntimeSIMDLib.cmake
@@ -1,3 +1,4 @@
+include(CheckCXXCompilerFlag)
 function(make_lib_simd_runtime name file)
     add_library(${name}_sse_or_arm STATIC)
     target_sources(${name}_sse_or_arm PRIVATE ${file})
@@ -6,10 +7,21 @@ function(make_lib_simd_runtime name file)
     target_sources(${name}_avx PRIVATE ${file})
     target_compile_definitions(${name}_avx PRIVATE BYOD_COMPILING_WITH_AVX=1)
     if(WIN32)
-        target_compile_options(${name}_avx PRIVATE /arch:AVX)
+        CHECK_CXX_COMPILER_FLAG("/arch:AVX" COMPILER_OPT_ARCH_AVX_SUPPORTED)
+        if(COMPILER_OPT_ARCH_AVX_SUPPORTED)
+            message(STATUS "Compiler supports flags: /arch:AVX")
+            target_compile_options(${name}_avx PRIVATE /arch:AVX)
+        else()
+            message(STATUS "Compiler DOES NOT supports flags: /arch:AVX")
+        endif()
     else()
-        # @TODO: can we exclude this for the ARM build altogether?
-        target_compile_options(${name}_avx PRIVATE -mavx -mfma -Wno-unused-command-line-argument)
+        CHECK_CXX_COMPILER_FLAG("-mavx -mfma" COMPILER_OPT_ARCH_AVX_SUPPORTED)
+        if(COMPILER_OPT_ARCH_AVX_SUPPORTED)
+            message(STATUS "Compiler supports flags: -mavx -mfma")
+            target_compile_options(${name}_avx PRIVATE -mavx -mfma -Wno-unused-command-line-argument)
+        else()
+            message(STATUS "Compiler DOES NOT supports flags: -mavx -mfma")
+        endif()
     endif()
 
     add_library(${name} INTERFACE)
